### PR TITLE
[FIRRTL] Add a pass to convert VoB -> BoV conversion

### PR DESCRIPTION
**Pull request by rwy7**
This pass converts objects with the shape `vector<bundle>` to `bundle<vector>`.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/VBToBV.cpp [hunk 1]
```diff
@@ -0,0 +1,988 @@
...
+  // Get the converted equivalent object.
+  value = valueMap[value];
+  assert(value && "canonical storage location must have been converted");
+
+  // Replay the subaccess operations, in the converted order;
+  // bundle accesses first, then vector accesses.
+  for (auto subfieldOp : llvm::reverse(bundleAccesses))
+    value = getSubfield(value, subfieldOp.getFieldIndex());
+
+  /// If the current value is a bundle, but we have vector accesses to replay,
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/VBToBV.cpp:475]: same

File: lib/Dialect/FIRRTL/Transforms/VBToBV.cpp [hunk 2]
```diff
@@ -0,0 +1,988 @@
...
+
+  if (value.getType().isa<BundleType>() && !vectorAccesses.empty()) {
+    explode(value, values);
+    exploded = true;
+  } else {
+    values.push_back(value);
+    exploded = false;
+  }
+
+  /// Finally, replay the vector access operations.
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/VBToBV.cpp:489]: here

File: lib/Dialect/FIRRTL/Transforms/VBToBV.cpp [hunk 3]
```diff
@@ -0,0 +1,988 @@
...
+  // with fresh intermediate bundle create ops in between.
+  SmallVector<Value> newOperands;
+  for (auto oldOperand : op->getOperands()) {
+    auto newOperand = fixROperand(oldOperand);
+    changed |= (oldOperand != newOperand);
+    newOperands.push_back(newOperand);
+  }
+
+  /// We can rewrite the type of any result, but if any result type changes,
+  /// then the operation will be cloned.
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/VBToBV.cpp:543]: here

**Review by youngar — APPROVED** [review]
LGTM,

**Review by uenoku — COMMENTED** [review]
Looks awesome! Does this work with bitcast op?

**Review by darthscsi — APPROVED** [review]
I'll leave the algorithmic review for others, but generally looks good.

File: lib/Dialect/FIRRTL/Transforms/VBToBV.cpp [hunk 4]
```diff
@@ -0,0 +1,988 @@
...
+          element.name, element.isFlip, convertType(element.type, dimensions)));
+    }
+    return BundleType::get(context, elements);
+  }
+  for (auto size : llvm::reverse(dimensions))
+    type = FVectorType::get(type, size);
+  return type;
+}
+
+FIRRTLBaseType Visitor::convertType(FIRRTLBaseType type) {
```
[the remark this answers is not in the document]
Comment (uenoku) [reply lib/Dialect/FIRRTL/Transforms/VBToBV.cpp:161]: I think `cached` could be invalid if `typeMap` is reallocated in the recursive calls.

[the remark this answers is not in the document]
Comment (darthscsi) [reply lib/Dialect/FIRRTL/Transforms/VBToBV.cpp:161]: Fair enough.  DenseMaps, sigh.

**Pull request merged**